doc/source/serve/production-guide/fault-tolerance.md

```diff
@@ -0,0 +1,304 @@
+The Serve controller logs the raised exception if the health-check fails.
+You can also use the deployment options to customize how frequently the health-check is run and the timeout after which a replica is marked unhealthy.
+
+```{literalinclude} doc_code/fault_tolerance/replica_health_check.py
+:start-after: __health_check_start__
+:end-before: __health_check_end__
+:language: python
+```
+
+### Worker node recovery
+
+:::{admonition} KubeRay Required
+:class: caution, dropdown
+You **must** deploy your Serve application with [KubeRay] to use this feature.
+
+See Serve's [Kubernetes production guide](serve-in-production-kubernetes) to learn how you can deploy your app with KubeRay.
+:::
+
+By default, Serve can recover from certain failures _within_ a Ray cluster, such as unhealthy actors. When [Serve runs on Kubernetes](serve-in-production-kubernetes) with [KubeRay], it can also recover from some _cluster-level_ failures, such as dead worker nodes.
+
+When a worker node fails, the actors running on it also fail. Serve detects that the actors have failed, and it attempts to respawn the actors on the remaining, healthy nodes. Meanwhile, KubeRay detects that the node itself has failed, so it brings up a new healthy node to replace it. Serve can then respawn any pending actors on that node as well. The deployment replicas running on healthy nodes can continue serving traffic throughout the recovery period.
+
+(serve-e2e-ft-guide-gcs)=
+### Head node recovery: Ray GCS fault tolerance
+
+:::{admonition} KubeRay Required
+:class: caution, dropdown
+You **must** deploy your Serve application with [KubeRay] to use this feature.
+
+See Serve's [Kubernetes production guide](serve-in-production-kubernetes) to learn how you can deploy your app with KubeRay.
+:::
+
+In this section, you'll learn how to add fault tolerance to Ray's Global Control Store (GCS), which allows your Serve application to serve traffic even when the head node crashes.
+
+By default the Ray head node is a single point of failure: if it crashes, the entire Ray cluster crashes and must be restarted. When running on Kubernetes, the `RayService` controller health-checks the Ray cluster and restarts it if this occurs, but this introduces some downtime.
+
+In Ray 2.0, KubeRay has added experimental support for [Global Control Store (GCS) fault tolerance](https://ray-project.github.io/kuberay/guidance/gcs-ft/#ray-gcs-fault-tolerancegcs-ft-experimental), preventing the Ray cluster from crashing if the head node goes down.
+While the head node is recovering, Serve applications can still handle traffic but cannot be updated or recover from other failures (e.g. actors or worker nodes crashing).
+Once the GCS is recovered, the cluster will return to normal behavior.
+
+You can enable GCS fault tolerance on KubeRay by adding an external Redis server and modifying your `RayService` Kubernetes object.
```

Comment:
```suggestion
You can enable GCS fault tolerance on KubeRay by adding an external Redis server and modifying your `RayService` Kubernetes object.
Below, we'll explain how to do each of these.
```
The section heading sizes are very similar so I thought this was the end of the section and got confused... I think this will make it a little bit smoother

Reply:
Good idea, I've updated the wording with your suggestion!